### PR TITLE
Ex1 layernotation

### PR DESCRIPTION
**Pull request by casperkaae**
Jeg har lavet lidt om i notationen i "exercises_1.ipynb" så det er konsistent ift implementering:
- $z^{(l)}$ is the input to a layer l
- $z^{(l+1)}$ is the output of layer l
- $t_j,...,t_k$  true targets
- $y_j,...,y_k$ predicted targets
- $h(\cdot)$ Nonlinear activation function (sigmoid, tanh, ReLU etc)

Forward pass er så 
- Linear layer: $z^{(l+1)} =  z^{(l)}w^l + b^{l}$ 
- Sigmoid layer: $z^{(l+1)} =  \sigma (z^{(l)} )$ 
- Softmax layer: $z^{(l+1)} =  \frac{exp(z^{(l)})}{\sum_j exp(z^{(l)}_j)}$ 

Backward pass:
$$ 
\delta^{(l)} = \frac{\partial E}{\partial z^{l}} = \frac{\partial E}{\partial z^{l+1}}\frac{\partial z^{l+1}}{\partial z^{l}} = \delta^{l+1}\frac{\partial z^{l+1}}{\partial z^{l}}^T 
$$
- Linear layer: $\delta^{(l)} =  \frac{\partial E}{\partial z^{l+1}}\frac{\partial z^{l+1}}{\partial z^l} = \frac{\partial (z^{(l)}w^{l+1} + b^{l+1})}{\partial z^l} = \delta^{(l+1)}w^{l} $
- Sigmoid layer: $\delta^{(l)} =  \frac{\partial E}{\partial z^{l+1}}\frac{\partial z^{l+1}}{\partial z^l} = \frac{\partial (\sigma(z^{l+1}))}{\partial z^l} = \delta^{(l+1)}\sigma'(z^{l}) $

Hvad tænker i om det? (åben evt filen for lidt flottere latex :) )

### CODE REVIEW

**Comment by casperkaae** [comment]
Jeg har tilføjet normalisering af cost med batch_størrelse samt nye implementeringer af softplus og tanh

**Comment by skaae** [comment]
Det ser godt ud.

I kan se det med ligninger på http://nbviewer.ipython.org/github/skaae/summerschool/blob/ex1_layernotation/exercises_1.ipynb

Smårettelser:   ( oprindelig -> rettet)
1. Let's start with a small toy network to get the not hold of the notation -> Let's start with a small toy network to get the hold of the notation
2. Vil du bruge y tilde eller bare y for netværks ssh?
    fx : predicts a target given an input x i.e.   (y tilde), men i bullet listen er det y (uden tilde)
3. Hence a layer is just a simple function parameterized w performing some transformation on its output. -> Hence a layer is just a simple function parameterized by w performing some transformation on its output.
4.  between theory and implementations ->  between theory and implementation
5. Specifically the forward pass is -> In a neural network with one hidden layer the forward pass is:
6. backpropation is tho calculate -> backpropation is to calculate
7. To calculate the loss function well simply plug in the output of the network into the loss function which "measures" the difference between the true and predicited targets.
8. The loss function takes the networks predictions and the true targets as input. It then "measures" the difference between the true and predicited targets.
9. which can be computed using the loss funtion -> which can be computed using the loss function
10. is the hadamard vector product i.e. element wise multiplications.        Hadamard bliver ikke brugt?
11. Specifically the forward pass is -> backward pass??

**Comment by casperkaae** [comment]
Du er meget velkommen til at comitte dine rettelser

**Comment by casperkaae** [comment]
Jeg har tilføjet rettelserne nu

**Comment by casperkaae** [comment]
Jeg har lavet lidt flere rettelser.

Har et enkelt spørgmål - Ved i hvorfor Test/Val Acc er bedre end Train acc?

**Comment by casperkaae** [comment]
Har kigget lidt mere på det - koden virker fint nok for et vanilla FFN. Problemet med training error der er dårligere end valid/test error er kun for dropout laget

**Comment by skaae** [comment]
Er det ikke bare fordi det skal disables under eval? Det ser ud til at der er en fejl i dropout implementeringen.

Det mest normale er at skallere output under træning. Jeg tror nok man bare ganger output under træning med 1/(1-p). Hvor p er ssh for at blive droppet. Dvs hvis p = 0.8 ganger man med 5. Under evaluering kan man så bare lade være med at bruge nogen skallering. 

``` Python
class DropoutLayer():
    def __str__(self): 
        return "Dropout(%f)" % self.p

    def __init__(self, p=0.5):
        self.p = p

    def fprop(self, x, train=True):
        if train:
            mask = np.random.random(x.shape) > (1-self.p)
            scale = 1.0 / (1-self.p)            
            self.a = x*mask*scale
            return self.a 
        else:
            return x 

    def bprop(self, delta_in):
        delta_out = delta_in*self.a
        return delta_out

    def update_params(self, lr):
        pass
```

**Comment by larsmaaloee** [comment]
Det er rigtigt Søren. Jeg ved ikke hvad for noget kode i snakker om her. Men det bliver vist også disabled under evaluering og test?
